### PR TITLE
 Block: Add story selector for embedding single story

### LINKED ISSUE #8845: Block: Add story selector for embedding single story
## Feature Description

Splitting this out from #6894

We want to make the Single Story embed block more similar to the image block, allowing to choose between selecting an existing story or inserting from URL:

Mock:

Screenshot 2021-08-12 at 16 07 51

The URL option would show the URL input field in a popup/tooltip like in the Image block.

The "Select Story" button would open the story selector we already have today, but only allow selecting 1 story (and not show a reorder option, obviously).

### PR DESCRIPTION
**Pull request by spacedmonkey**
## Context

We want to make the Single Story embed block more similar to the image block, allowing to choose between selecting an existing story or inserting from URL:

Mock:

Screenshot 2021-08-12 at 16 07 51

The URL option would show the URL input field in a popup/tooltip like in the Image block.

The "Select Story" button would open the story selector we already have today, but only allow selecting 1 story (and not show a reorder option, obviously).

## Summary

Make the Single Story embed block more similar to the image block.

## Relevant Technical Choices

- Updated PHP logic, new story is created, the id of the story is embedded. This means in future, we have this data stored. It also means that user can edit the story block. 
- Update E2E and Unit tests. 
- Moved all the story picker and related files into share components from there current location in selected stories block type.  
- Rename Story url to story embed. 
- Make `StoryPicker` much generic, so it can be used in both places. 

#### New interface for selecting stories. 

https://user-images.githubusercontent.com/237508/134694059-27e991dc-e2c5-4747-a6e1-0b18c3a73d1d.mov

#### New interface for embedding via URL.

https://user-images.githubusercontent.com/237508/134694097-7e5eb37c-b015-4e54-9005-fdb2c1ef62db.mov

## Testing Instructions

- [ ] This is a non-user-facing change and requires no QA

This PR can be tested by following these steps:

#### Embed story via selector

1. Create a new post. 
3. Insert a web story block. 
4. Click on "Single Story". 
5. Click on "Select Story". 
6. Select a story. 
7. See embedded story. 
8. Click on "Replace" in block toolbar, see "Replace Story" button 
9. Select another story. 
10. See embedded story has changed. 

#### Embed story via URL

1. Create a new post. 
3. Insert a web story block. 
4. Click on "Single Story". 
5. Click insert from url. 
6. Enter a url for a story, for example https://wp.stories.google/stories/how-to-optimize-videos-for-web-stories/
7. See embedded story. 
8. Click on "Replace" in block toolbar, see "Replace URL" button
9. Enter a URL for a story. 
10. See embedded story has changed.

### Does this PR have a security-related impact?

No

### Does this PR change what data or activity we track or use?

No

### Does this PR have a legal-related impact?

No

## Checklist

- [x] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [x] I have tested this code to the best of my abilities
- [x] I have verified accessibility to the best of my abilities ([docs](https://github.com/google/web-stories-wpx
- [x] /blob/main/docs/accessibility-testing.md))
- [x] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [x] This code is covered by automated tests (unit, integration, and/or e2e) to verify it works as intended ([docs](https://github.com/google/web-stories-wp/tree/main/docs#testing))
- [x] I have added documentation where necessary
- [x] I have added a matching `Type: XYZ` label to the PR

---

Fixes #8845

### CODE REVIEW

File: includes/Admin/Admin.php [hunk 1]
```diff
@@ -156,7 +156,7 @@ public function prefill_post_content( $content, $post ) {
 		$renderer = new Image( $story );
 		$html     = $renderer->render( $args );
 
-		$content = '<!-- wp:web-stories/embed {"url":"%1$s","title":"%2$s","poster":"%3$s","width":"%4$s","height":"%5$s","align":"%6$s"} -->%7$s<!-- /wp:web-stories/embed -->';
+		$content = '<!-- wp:web-stories/embed {"blockType":"url","url":"%1$s","title":"%2$s","poster":"%3$s","width":"%4$s","height":"%5$s","align":"%6$s","stories": [%7$s]} -->%8$s<!-- /wp:web-stories/embed -->';
```
Comment (spacedmonkey) [inline includes/Admin/Admin.php:159]: The block type here is a fix we should have done a while ago.

File: packages/stories-block/src/block/block-types/latest-stories/edit.js [hunk 2]
```diff
@@ -70,7 +70,7 @@ function LatestStoriesEdit({ attributes, setAttributes }) {
       const stories = await apiFetch({
         path: addQueryArgs(storiesApi, {
           per_page: 20,
-          _embed: 'author',
+          _embed: 'author,wp:featuredmedia',
```
Comment (spacedmonkey) [inline packages/stories-block/src/block/block-types/latest-stories/edit.js:73]: Unrelated fix

File: packages/stories-block/src/block/block-types/single-story/edit.js [hunk 3]
```diff
@@ -107,9 +126,7 @@ function StoryEmbedEdit({
 
         setCannotEmbed(!(typeof data?.title === 'string'));
         setStoryData(data);
-        setAttributes({
-          url: localURL,
-        });
+        setAttributes({ url });
```
Comment (spacedmonkey) [inline packages/stories-block/src/block/block-types/single-story/edit.js]: We didn't need to use `localURL` here, as it will be the same value of url. It better for the callback.

File: packages/stories-block/src/block/block-types/single-story/embedPlaceholder.js [hunk 4]
```diff
@@ -34,58 +65,92 @@ const EmbedPlaceholder = ({
...
+          'Select an existing story, or add one with a URL.',
+          'web-stories'
+        )}
+      >
+        <>
+          <Button isPrimary onClick={openStoryPicker}>
+            {!hasStories
+              ? __('Select Story', 'web-stories')
+              : __('Replace Story', 'web-stories')}
+          </Button>
```
Comment (spacedmonkey) [inline packages/stories-block/src/block/block-types/single-story/embedPlaceholder.js:86]: Does this need to be wrapped in a form?

File: packages/stories-block/src/block/components/storyPicker/storyPicker.js [hunk 5]
```diff
@@ -157,25 +163,26 @@ function StoryPicker({
...
-            </Button>
-          ) : (
-            <Button
-              onClick={() => setIsSortingStories(true)}
-              disabled={localSelectedStories.length < 2}
-            >
-              {__('Rearrange Stories', 'web-stories')}
-            </Button>
-          )}
+          {maxNumOfStories > 1 &&
```
Comment (spacedmonkey) [inline packages/stories-block/src/block/components/storyPicker/storyPicker.js:173]: Only show rearrange button, if max number of stories in more than 1.

File: packages/stories-block/src/block/components/storyPicker/storyPicker.js [hunk 6]
```diff
@@ -119,7 +119,12 @@ function StoryPicker({
 
   return (
     <Modal
-      title={__('Selected Stories', 'web-stories')}
+      title={_n(
+        'Selected Story',
+        'Selected Stories',
+        maxNumOfStories,
+        'web-stories'
+      )}
```
Comment (swissspidy) [inline packages/stories-block/src/block/components/storyPicker/storyPicker.js]: `_n()` should only be used if the string contains a number/placeholder.

If not, need to do `1 === maxNumOfStories ? _n('Selected Story', 'web-stories') : _n('Selected Stories', 'web-stories')`

File: packages/stories-block/src/block/block-types/latest-stories/edit.js (hunk 2 above)
Comment (swissspidy) [reply packages/stories-block/src/block/block-types/latest-stories/edit.js:73]: Curious, what was broken / what does this fix?

Comment (spacedmonkey) [reply packages/stories-block/src/block/block-types/latest-stories/edit.js:73]: The poster image was not displaying as it was not loaded in the API.

**Comment by swissspidy** [comment]
There are some e2e test failures due to React warnings and probably some bugs

**Comment by spacedmonkey** [comment]
> There are some e2e test failures due to React warnings and probably some bugs

E2E tests are now passing. Super simple fix!

File: packages/stories-block/src/block/block-types/single-story/embedPlaceholder.js [hunk 7]
```diff
@@ -22,9 +22,41 @@ import PropTypes from 'prop-types';
...
+import { Button, Placeholder } from '@wordpress/components';
+import { BlockIcon, URLPopover } from '@wordpress/block-editor';
+import { __, sprintf } from '@wordpress/i18n';
+import { keyboardReturn } from '@wordpress/icons';
+/**
+ * Internal dependencies
+ */
+import StoryPicker from '../../components/storyPicker/storyPicker';
+
+// eslint-disable-next-line react/prop-types
```
Comment (swissspidy) [inline packages/stories-block/src/block/block-types/single-story/embedPlaceholder.js]: Why not just add some prop types?

Comment (spacedmonkey) [reply packages/stories-block/src/block/block-types/single-story/embedPlaceholder.js]: Fixed

File: packages/stories-block/src/block/block-types/single-story/edit.js [hunk 8]
```diff
@@ -14,6 +14,8 @@
  * limitations under the License.
  */
 
+/* eslint complexity: ["error", { "max": 21 }] */
```
Comment (swissspidy) [inline packages/stories-block/src/block/block-types/single-story/edit.js:17]: Sounds like it's time for some refactoring soon to fix this instead of just bumping the number here

**Review by swissspidy — APPROVED** [review]
Made some minor wording tweaks as discussed, otherwise looks good and works great!
